p2p/net/swarm/swarm.go

```diff
@@ -412,10 +418,10 @@ func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn,
 
 	// Emit event after releasing `s.conns` lock so that a consumer can still
 	// use swarm methods that need the `s.conns` lock.
-	if isFirstConnection {
+	if oldState != newState {
 		s.emitter.Emit(event.EvtPeerConnectednessChanged{
 			Peer:          p,
-			Connectedness: network.Connected,
+			Connectedness: newState,
 		})
```

Comment:
You need the same changes here on the disconnection side: https://github.com/libp2p/go-libp2p/blob/57736f1ec68de8471650a65ff383cb0ada0a0c4d/p2p/net/swarm/swarm.go#L777

Reply:
What checks exactly are required?

IIUC it isn't necessary to delete the conn from `s.conns.m` ?

Reply:
maintaining the conns map is okay, my comment is regarding sending the `PeerConnectednessEvent` when a connection is disconnected. 
https://github.com/libp2p/go-libp2p/blob/57736f1ec68de8471650a65ff383cb0ada0a0c4d/p2p/net/swarm/swarm.go#L777-L787
These lines should now check whether the connection we have to the peer is a Limited connection or an Unlimited connection and send a connectedness change event appropriately.